feat(typescript): add declaration files (#4220)
algolia/instantsearch.js: CONTRIBUTING.md

```

_Make sure to use `npm run` instead of `yarn run` to avoid issues._

#### Experimental TypeScript version

> An experimental version containing the TypeScript declaration files is available.

Reviewer: ```suggestion
An experimental version containing the TypeScript declaration files is available on the npm tag `experimental-typescript`.
```

> Since some of these declaration files are generated from the JSDoc comments they can still contains some typing errors. This version will stay experimental until we are confident enough in the generated declarations to put them in the stable release.

Reviewer: ```suggestion
Since some of these declaration files are generated from the JSDoc comments, they can contain some typing errors. This version will stay experimental until we are confident enough in the generated declarations to put them in a stable release.
```

> To generate the experimental TypeScript version for a particular release run:

Reviewer: ```suggestion
To generate the experimental TypeScript version for a particular release, run:
```

```sh
git checkout v4.X.X
VERSION=4.X.X-experimental-typescript.X node ./scripts/release/bump-package-version.js
yarn build
yarn build:types
sed -i '/"main":/ a \ "types": "es/index.d.ts",' package.json # Add `types` entry in `package.json`
npm publish --tag experimental-typescript
git checkout .

> ```

Reviewer: Would it make sense to also push the code of the release on a GitHub tag? This will allow us to track what's been released.

Author: I don't think so. Since we will not push the changes for `experimental-typescript` to GitHub, the tag would always be on the same commit as for the stable one (ex: on `v4.X.X` in the script above).

Reviewer: Are you reluctant to this because it's more work to do on our side or because it doesn't provide value?

Ideally, you should be able to map an npm release with a GitHub tag. If we release without publishing the source code related to this release, it's like we "blindly" release without having the ability to check (even later), what this release contained.

I see this as an improvement (that could be handled by Ship.js later), not as a blocker.

Author: FMPOV it just doesn't provide value.

The source code for `v4.X.X-experimental-typescript.X` will be the same as the one of `v4.X.X` (minus the `version` and `types` changes in `package.json` that are not committed) so no need for a new tag.
